### ISSUE TITLE
Add new single HTML radio button component that isn't bound to a list

### ISSUE THREAD

**Issue by glassfishrobot** [javaee/javaserverfaces-spec#329]
The selectOneRadio component requires that all radio buttons be grouped
together. This is very inflexible when creating user interfaces. Many times I
have encountered screens where I need multiple radio buttons in the same group,
but each radio button exists in a separate area of the screen. This can be done
in HTML easily, but not with JSF's default component set.

I've read that Apache MyFaces has created a component to solve this issue, and
so has Sun's Woodstock component set. Please standardize this fundamental
component in JSF 2.0 so that all implementations come with it.
#### Environment
Operating System: All
Platform: All
#### Affected Versions
[2.0]

**Comment by glassfishrobot** [comment]
Reported by rdelaplante

**Comment by glassfishrobot** [comment]
rdelaplante said:
Created an attachment (id=122)
Example form where this component would be useful

**Comment by glassfishrobot** [comment]
File: [radio buttons.png](https://java.net/jira/secure/attachment/13649/radio buttons.png)
Attached By: rdelaplante

**Comment by glassfishrobot** [comment]
@edburns said:
Per 20080827 EG Meeting, push to 2.1

**Comment by glassfishrobot** [comment]
@edburns said:
Very early in the development of JSF 2.0, the EG decided we didn't have enough
resources to add new components and that, since components are extensible
anyway, our time would be better spent working on the top five features of 1.
making components easier to develop 2\. having first class ajax support for
custom components 3\. reducing the configuration burden 4\. providing better error
handling and developer experience 5\. providing for better interoperability
between 3rd party jsf components.

**Comment by glassfishrobot** [comment]
rdelaplante said:
From my experience, this is the only missing component that maps directly to
HTML. I think this is different from adding fun and useful new components.

In my web apps that don't use sophisticated components such as calendars, I
prefer to use the basic h:* input components so that I don't have to force
hundreds of KB downloads on my users. I style with my own CSS. Since JSF
doesn't offer a radio button component where I can specify the html group name
as an attribute, I have to introduce hundreds of KB download of CSS and
JavaScript to my users by adding a component suite.

**Comment by glassfishrobot** [comment]
rdelaplante said:
This was most noticeable when creating a slim UI for mobile web browsers. If I
wanted to create a UI such as the one in the attached screenshot, mobile users
(smart phones on slow networks) would have to wait 2+ minutes while the
component suite's CSS and JavaScript would download. If JSF had a built in
radio button component that maps to HTML, then this wouldn't be a problem.

**Comment by glassfishrobot** [comment]
rdelaplante said:
I have found a solution for the problem outlined in this ticket without
requiring the addition of a new component. I noticed that issue #229 has added
two new attributes to selectManyCheckbox. All that is missing from
selectOneRadio is a name attribute.

<h:selectOneRadio id="notifyOff" value="OFF" name="notifyType"/>
<h:selectOneRadio id="notifyEmail" value="EMAIL" name="notifyType"/>
<h:selectOneRadio id="notifySMS" value="SMS" name="notifyType"/>

The name attribute is what makes the Woodstock Component Suite's radio button
usable for me, because it lets me have control of the screen layout as shown in
the attached screenshot. If you can add a name attribute to the 2.0 spec that
would be really great!

**Comment by glassfishrobot** [comment]
rdelaplante said:
I meant to refer to issue #228 instead of #229

**Comment by glassfishrobot** [comment]
rdelaplante said:
Here are some examples of how developers have solved this issue:

[http://webdev2.sun.com/woodstock-tlddocs/webuijsf/radioButton.html](http://webdev2.sun.com/woodstock-tlddocs/webuijsf/radioButton.html)
[http://www.icefaces.org/docs/v1_7_2sp1/tld/ice/radio.html](http://www.icefaces.org/docs/v1_7_2sp1/tld/ice/radio.html)
[http://myfaces.apache.org/tomahawk-project/tomahawk12/tlddoc/t/radio.html](http://myfaces.apache.org/tomahawk-project/tomahawk12/tlddoc/t/radio.html)
[http://software.topcoder.com/catalog/c_component.jsp?comp=26817861&ver=1](http://software.topcoder.com/catalog/c_component.jsp?comp=26817861&ver=1)

**Comment by glassfishrobot** [comment]
@edburns said:
Prepare to delete "spec" subcomponent.

**Comment by glassfishrobot** [comment]
@edburns said:
Move these to unscheduled because we need to target them correctly. 2.next isn't
specific enough.

**Comment by glassfishrobot** [comment]
rogerk said:
cat2

**Comment by glassfishrobot** [comment]
@edburns said:
renderkitdoc

**Comment by glassfishrobot** [comment]
@edburns said:
These are targeted at 2.1.

**Comment by glassfishrobot** [comment]
sheetalv said:
triage

**Comment by glassfishrobot** [comment]
@edburns said:
rogerk

**Comment by glassfishrobot** [comment]
rogerk said:
triage

**Comment by glassfishrobot** [comment]
rogerk said:
triage

**Comment by glassfishrobot** [comment]
rdelaplante said:
I like how the Trinidad and IceFaces radio button works. Essentially what they've done is extend the existing h:selectOneRadio and added a new layout constant value called "spread" which tells it to not render itself. Then they created a new t:radio component that renders a single radio button and its label wherever you place it. The t:radio has a "for" attribute that points back to the main selectOneRadio, and an "index" attribute that tells it which of the select items to output. It works with JSF's built-in AJAX and I love it. My only issue is that I can't use it in a ui:repeat loop, but I suspect that is a JSF spec issue related to findComponent. I'd also like to be able to choose if the label is rendered to the left or right of the radio button, or not render the label at all.

<t:selectOneRadio id="upgradesRadio" layout="spread" value="#

{model.upgradeSelection}

">
<f:selectItems value="#

{model.availableUpgrades}

" var="upgrade" itemLabel="#

{upgrade.description}

"/>
<f:ajax/>
</t:selectOneRadio>

... custom HTML layout ...

<t:radio for="upgradesRadio" index="0"/>

... custom HTML layout ...

<t:radio for="upgradesRadio" index="1"/>

**Comment by glassfishrobot** [comment]
rdelaplante said:
Please PLEASE get this done in JSF 2.3, we've been waiting way too long for this. With a perfect solution in Trinidad that works well with the existing h:selectOneRadio design, I don' understand why this keeps getting left undone release after release. Search Google for people using JSF and radio buttons and you'll find a lot of misery.

**Comment by glassfishrobot** [comment]
rdelaplante said:
I was working with the t:radio component from Trinidad recently and discovered an inflexibility. It always renders a label with the radio button and so I had difficulty implementing the layout and wrapping dictated by a web designer. Also, they wanted part of the label text to have different font styling. I was unable to embed any kind of HTML tags in the label text loaded from the backing bean (such as a span or div) because it always escapes the label text.

If you implement this feature in JSF 2.3 (please do!), it would be nice to have the option to output just the radio button without the label, and also the option to not escape the text in the label.

**Comment by glassfishrobot** [comment]
@edburns said:
Set priority to baseline ahead of JSF 2.3 triage. Priorities will be assigned accurately after this exercise.

**Comment by glassfishrobot** [comment]
@manfredriem said:
Setting priority to Critical

**Comment by glassfishrobot** [comment]
rdelaplante said:
Another consideration for this component is the ability to bind the value to a Boolean. I've had numerous customers provide screen layouts with two radio buttons and specific wording that essentially boil down to yes/no or on/off, and I end up changing the wording to make it work with a checkbox instead. Some customers don't like that so I have to write backing bean code to read the value from the radio button component and store it in my model manually.

**Comment by glassfishrobot** [comment]
rdelaplante said:
I'm monitoring the thread about this ticket on the JSF expert group mailing list but don't think I can reply to it so I'm posting my comments here. Specifically I'm commenting on Manfred's post about it being simple enough to register your own renderer and recommending that this ticket be closed as "Won't Fix". Also on Manfred's remark about a new component being overkill.

I created this ticket seven years ago and wrote most of the comments. I no longer recommend my original solution from the earliest comments. Instead I recommend the Tomahawk solution where they added a new value to the h:selectOneRadio layout attribute called "spread" which tells it to not render anything. They also created a new t:radio component which has a for attribute that points to the h:selectOneRadio component's id, and an index attribute to specify which SelectItem to use. IceFaces implemented the same solution.

Two component suites implemented the same solution which work well for me in my applications. It sounds like Manfred has a simpler idea. Can you please describe it?

As a developer who has been using JSF to create web applications for at least seven years now, I think that I am qualified to say that JSF does not currently and never has supported HTML radio buttons. The h:selectOneRadio component can only be used in trivial use cases which happens to be never, not even once, in all the years I've been using JSF. If it weren't for the Tomahawk and IceFaces solutions (crutches IMO) then developers like myself would not have been able to use JSF in our applications and wouldn't still be using JSF today.

Ed has asked me to volunteer to implement this. I've agreed to give it a shot but I really don't know the source code and would very much prefer that a real expert do it.

**Comment by glassfishrobot** [comment]
@BalusC said:
I'll take it on me.

**Comment by glassfishrobot** [comment]
@BalusC said:
Summarized: best approach would be to modify the existing renderer in such way so that at least the below cases are possible:

1:

```
<h:dataTable value="#{bean.items}" var="item">
    <h:column>
        <h:selectOneRadio group="foo" value="#{bean.selectedItem}">
            <f:selectItem itemValue="#{item}" />
        </h:selectOneRadio>
    </h:column>
</h:dataTable>
```

2:

```
<ui:repeat value="#{bean.items}" var="item">
    <h:outputLabel for="radioId" value="#{item.label}" />
    <h:selectOneRadio id="radioId" group="foo" value="#{bean.selectedItem}">
        <f:selectItem itemValue="#{item.value}" />
    </h:selectOneRadio>
    <br/>
</ui:repeat>
```

3:

```
<h:selectOneRadio group="foo" value="#{bean.selectedItem}">
    <f:selectItem itemValue="one" />
    <f:selectItem itemValue="two" />
    <f:selectItem itemValue="three" />
</h:selectOneRadio>
<h:selectOneRadio group="foo" />
<h:selectOneRadio group="foo" />
```

4:

```
<h:selectOneRadio group="foo" value="#{bean.selectedItem}">
    <f:selectItem itemValue="one" />
</h:selectOneRadio>
<h:selectOneRadio group="foo">
    <f:selectItem itemValue="two" />
</h:selectOneRadio>
<h:selectOneRadio group="foo">
    <f:selectItem itemValue="three" />
</h:selectOneRadio>
```

5:

```
<h:selectOneRadio group="foo" value="#{bean.selectedItem}">
    <f:selectItems value="#{['one', 'two', 'three']}" />
</h:selectOneRadio>
<h:selectOneRadio group="foo" />
<h:selectOneRadio group="foo" />
```

The <tt>group</tt> attribute is new. When <tt>group</tt> is specified, it must render solely a <tt><input type="radio"></tt> element (no table, no list). The <tt>group</tt> value must be relative to the <tt>UIForm</tt> and represent the common HTML name.

If there are multiple components with same <tt>group</tt> in the <tt>UIForm</tt>, and the <tt>value</tt> attribute and/or <tt>UISelectItem</tt> child is absent, then the one of the first component must be considered. The <tt>UISelectItem</tt> values will be assigned to the components having same <tt>group</tt> in the same order as components appear in the <tt>UIForm</tt>.

If the <tt>UISelectItem</tt> has a item label, then a <tt><label></tt> will be rendered directly _after_ the <tt><input></tt> element, without any additional markup.

**Comment by glassfishrobot** [comment]
kito75 said:
+1 for this approach

**Comment by glassfishrobot** [comment]
@BalusC said:
[https://java.net/projects/mojarra/sources/git/revision/1edee483ce009c929d11f8c5a53dc11688c10cf5](https://java.net/projects/mojarra/sources/git/revision/1edee483ce009c929d11f8c5a53dc11688c10cf5)

**Comment by glassfishrobot** [comment]
Marked as **fixed** on Wednesday, September 28th 2016, 4:03:47 am

**Comment by glassfishrobot** [comment]
@BalusC said:
Implementation follows in [https://java.net/jira/browse/JAVASERVERFACES-4188](https://java.net/jira/browse/JAVASERVERFACES-4188)

**Comment by glassfishrobot** [comment]
@BalusC said:
When [https://java.net/jira/browse/JAVASERVERFACES_SPEC_PUBLIC-1433](https://java.net/jira/browse/JAVASERVERFACES_SPEC_PUBLIC-1433) is enabled, the <h:selectOneRadio group="..." required="true"> displays multiple required message, one per component, instead of only once. Due to this, UISelectOne#processValidators() needed to be altered to cover this case, which is done in this commit: [https://java.net/projects/mojarra/sources/git/revision/554d49e57c62143287521cef879b3444601cca5a](https://java.net/projects/mojarra/sources/git/revision/554d49e57c62143287521cef879b3444601cca5a)

Improved implementation and IT follows in [https://java.net/jira/browse/JAVASERVERFACES-4188](https://java.net/jira/browse/JAVASERVERFACES-4188)

**Comment by glassfishrobot** [comment]
This issue was imported from java.net JIRA JAVASERVERFACES_SPEC_PUBLIC-329

**Comment by cletecillc** [comment]
When using `h:selectOneRadio` with `group` attribute and <f:ajax listener="...."/>, the listener method isn't called. Ex:
```
<h:selectOneRadio id="dpToday" group="dpFilter"  value="#{bean.today}">
    <f:selectItem itemValue="0" />
    <f:ajax event="change" listener="#{bean.method}"/>
</h:selectOneRadio>
```
The method `ben.method` is never called.
